src/sim/syscall_abi.hh

```diff
@@ -78,19 +79,22 @@ struct GenericSyscallABI32 : public GenericSyscallABI
 namespace guest_abi
 {
 
+using namespace pseudo_inst;
+
 // For 64 bit systems, return syscall args directly.
 template <typename ABI, typename Arg>
 struct Argument<ABI, Arg,
     typename std::enable_if_t<
         std::is_base_of_v<GenericSyscallABI64, ABI> &&
-        std::is_integral_v<Arg>>>
+        (std::is_integral_v<Arg> || std::is_same<Arg,GuestAddr>::value)>>
 {
     static Arg
     get(ThreadContext *tc, typename ABI::State &state)
     {
         panic_if(state >= ABI::ArgumentRegs.size(),
                 "Ran out of syscall argument registers.");
-        return tc->getReg(ABI::ArgumentRegs[state++]);
+        auto arg = tc->getReg(ABI::ArgumentRegs[state++]);
+        return *reinterpret_cast<Arg*>(&arg);
```

Comment:
This is ok, but I would honestly pay the price of duplication and treat is_integral and GuestAddr separately, so that you are not forced to use the reinterpret_cast trick

Reply:
The reinterpret_cast would be still necessary in the get function for GuestAddr, because it is not possible to cast uint64_t to GuestAddr. If you are OK with it, i would like to keep both cases in one function.

Reply:
I don't have super strong objection against it, I am just trying to avoid reinterpret casting as much as possible in cases where this is not needed.

Unless I am missing something you don't need casting for GuestAddr if you do:

        auto arg = tc->getReg(ABI::ArgumentRegs[state++]);
        return GuestAddr{arg};

Alternatively you can avoid casting by providing conversion operators(ish) to the GuestAddr. I didn't check but I believe that if you do

        struct GuestAddr
        {
            GuestAddr(Addr _addr) : addr(_addr) {}
            Addr addr;     
        };

Doing the following should just work:

       retrurn tc->getReg(ABI::ArgumentRegs[state++]);